src/Temporalio.Extensions.Hosting/ServiceProviderExtensions.cs

```diff
@@ -97,5 +101,66 @@ public static class ServiceProviderExtensions
             };
             return ActivityDefinition.Create(method, invoker);
         }
+
+#if NET6_0_OR_GREATER
+        /// <summary>
+        /// Create activity definition for the given activity-attributed method on the given
+        /// instance type. If the method is non-static, this will use the service provider to get
+        /// the instance to call the method on.
+        /// </summary>
+        /// <param name="provider">Service provider for creating the instance if the method is
+        /// non-static.</param>
+        /// <param name="instanceType">Type of the instance.</param>
+        /// <param name="method">Method to create activity definition from.</param>
+        /// <returns>Created definition.</returns>
+        public static ActivityDefinition AsyncCreateTemporalActivityDefinition(
```

Comment:
I think it'd be better to just alter the existing public method to support both .NET versions instead of someone having to know which .NET version of this they want if they wanted to call this public API directly.

I know it can make the code a little annoying, but can we alter the original method to work with both instead of making a new method? I think the easiest approach may be a `try`/`finally` instead of `using`. Can discuss potential private static helpers if the `try`/`finally` with inline compiler directives is not ok.

Reply:
That's a good point, I understand it. With the way this is configured, only .NET8 would see two options which is the route I went let me see what I can come up with for cleanliness.

Reply:
:+1: So local functions is still supported in our compiled versions, so that change works. And to be able to conditionally choose between `await using` and just `using` I think means not to use `using` but `try`/`finally` instead.

Reply:
How about using Public with PrivateCore pattern like the .NET runtime?

Example.)
```csharp
/// <summary>
/// Create activity definitions for every activity-attributed method on the given type. For
/// non-static methods, this will use the service provider to get the instance to call the
/// method on.
/// </summary>
/// <param name="provider">Service provider for creating the instance for non-static
/// activities.</param>
/// <param name="type">Type to create activity definitions from.</param>
/// <returns>Collection of activity definitions.</returns>
public static IReadOnlyCollection<ActivityDefinition> CreateTemporalActivityDefinitions(
    this IServiceProvider provider, Type type) =>
    type.
        GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.Instance).
        Where(method => method.IsDefined(typeof(ActivityAttribute))).
        Select(method => provider.CreateTemporalActivityDefinition(type, method)).
        ToList();

/// <summary>
/// Create activity definition for the given activity-attributed method on the given
/// instance type. If the method is non-static, this will use the service provider to get
/// the instance to call the method on.
/// </summary>
/// <param name="provider">Service provider for creating the instance if the method is
/// non-static.</param>
/// <param name="instanceType">Type of the instance.</param>
/// <param name="method">Method to create activity definition from.</param>
/// <returns>Created definition.</returns>
public static ActivityDefinition CreateTemporalActivityDefinition(
    this IServiceProvider provider, Type instanceType, MethodInfo method)
{
#if NET6_0_OR_GREATER
    return provider.AsyncCreateTemporalActivityDefinitionCore(instanceType, method);
#else
    return provider.CreateTemporalActivityDefinitionCore(instanceType, method);
#endif
}
```

Reply:
I would work and I'd be ok with it, but I think you'd just be duplicating lots of the lines unnecessarily. I think this is clear:

```csharp
        public static ActivityDefinition CreateTemporalActivityDefinition(
            this IServiceProvider provider,
            Type instanceType,
            MethodInfo method)
        {
            // Invoker can be async (i.e. returns Task<object?>)
            async Task<object?> Invoker(object?[] args)
            {
                // Wrap in a scope (even for statics to keep logic simple)
#if NET6_0_OR_GREATER
                var scope = provider.CreateAsyncScope();
#pragma warning restore CA2007
#else
                var scope = provider.CreateScope();
#endif

                try
                {
                    object? result = null;
                    try
                    {
                        // Invoke static or non-static
                        var instance = method.IsStatic
                            ? null
                            : scope.ServiceProvider.GetRequiredService(instanceType);

                        result = method.Invoke(instance, args);
                    }
                    catch (TargetInvocationException e)
                    {
                        ExceptionDispatchInfo.Capture(e.InnerException!).Throw();
                    }

                    // In order to make sure the scope lasts the life of the activity, we need to
                    // wait on the task if it's a task
                    if (result is Task resultTask)
                    {
                        await resultTask.ConfigureAwait(false);
                        // We have to use reflection to extract value if it's a Task<>
                        var resultTaskType = resultTask.GetType();
                        if (resultTaskType.IsGenericType)
                        {
                            result = resultTaskType.GetProperty("Result")!.GetValue(resultTask);
                        }
                        else
                        {
                            result = ValueTuple.Create();
                        }
                    }
                    return result!;
                }
                finally
                {
#if NET6_0_OR_GREATER
                    await scope.DisposeAsync();
#else
                    scope.Dispose();
#endif
                }
            }
            return ActivityDefinition.Create(method, Invoker);
        }
```

Reply:
Understood, I wasn't sure if you were keen on the conversion to local function in both was my only concern. I will submit this change.